### ISSUE TITLE
Nicer cache invalidation

### ISSUE THREAD

**Issue by guybedford** [systemjs/builder#328]
Seeing https://github.com/jackfranklin/jspm-dev-builder/blob/master/index.js#L13 this is definitely fiddly to put together. Perhaps we can do something nicer here.

If the cache was itself a class, we could have invalidation methods on it like:

```javascript
var cache = new Builder.Cache();

builder.setCache(cache);
// do build stuff
cache.invalidate('some/file.js') // automatically invalidates everything relying on the source of this file
cache.inject('some/file.js', source); // set the source of a file programatically
cache.toString() // get the serialized cache representation

new Builder.Cache(cache.toString()) // deserialize cache strings
```

Just some random ideas.

//cc @jackfranklin

**Comment by jackfranklin** [comment]
@guybedford this looks great - pinging @oliverjash who I was also chatting this through with. You're right in saying it's pretty fiddly right now, and something like this would really help with that.

**Comment by guybedford** [comment]
I'm wondering if we should just make this `builder.invalidate(moduleName)` rather than a special cache object, which can then act on the caches in the right ways.

**Comment by jackfranklin** [comment]
:+1: the tricky bit was the fact that the name in the trace cache isn't just the file name - it would be good to not have to have knowledge of that to be able to invalidate a file. Feels like leaky internals.

**Comment by guybedford** [comment]
I've created a PR in https://github.com/systemjs/builder/pull/341.

**Comment by guybedford** [comment]
Released in 0.14.8.
